web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx

```diff
@@ -0,0 +1,269 @@
+                <Icons.ChevronRight
+                  size="small"
+                  color="text.muted"
+                  css={`
+                    right: 8px;
+                    position: absolute;
+                  `}
+                />
+              </Flex>
+              <Text color={colors.client}>{latency.client}ms</Text>
+            </Flex>
+
+            <Flex
+              gap={1}
+              width="24px"
+              flexDirection="column"
+              alignItems="center"
+              css={`
+                flex-grow: 0;
+              `}
+            >
+              <TeleportGearIcon size={24}></TeleportGearIcon>
+              <Text>Teleport</Text>
+            </Flex>
+
+            <Flex
+              gap={1}
+              flexDirection="column"
+              alignItems="center"
+              css={`
+                flex-grow: 1;
+                position: relative;
+              `}
+            >
+              <Flex
+                gap={1}
+                flexDirection="row"
+                alignItems="center"
+                width="100%"
+                css={`
+                  padding: 0 16px;
+                `}
+              >
+                <Icons.ChevronLeft
+                  size="small"
+                  color="text.muted"
+                  css={`
+                    left: 8px;
+                    position: absolute;
+                  `}
+                />
+                <Line />
+                <Icons.ChevronRight
+                  size="small"
+                  color="text.muted"
+                  css={`
+                    right: 8px;
+                    position: absolute;
+                  `}
+                />
+              </Flex>
+              <Text color={colors.server}>{latency.server}ms</Text>
+            </Flex>
+
+            <Flex
+              gap={1}
+              width="24px"
+              flexDirection="column"
+              alignItems="flex-end"
+              css={`
+                flex-grow: 0;
+              `}
+            >
+              <Icons.Server />
+              <Text>Server</Text>
+            </Flex>
+          </Flex>
+
+          <Flex flexDirection="column" alignItems="center">
+            <Flex gap={1} flexDirection="row" alignItems="center">
+              {colors.total === latencyColor.error && (
+                <Icons.WarningCircle size={20} color={colors.total} />
+              )}
+
+              {colors.total === latencyColor.warn && (
+                <Icons.Warning size={20} color={colors.total} />
+              )}
+
+              {colors.total === latencyColor.ok && (
+                <Icons.CircleCheck size={20} color={colors.total} />
+              )}
+
+              <Text bold fontSize={2} textAlign="center" color={colors.total}>
+                Total Latency: {latency.total}ms
+              </Text>
+            </Flex>
+          </Flex>
+        </Flex>
+      </Container>
+    </MenuIcon>
+  );
+}
```

Comment:
The CSS here is generally fine, my critique is that there are several repeated component hierarchies / patterns that can be parameterized. I did so locally, here's the patch:

<details>

```patch
diff --git a/web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx b/web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx
index 8b5889038a..87e2bd5b93 100644
--- a/web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx
+++ b/web/packages/shared/components/LatencyDiagnostic/LatencyDiagnostic.tsx
@@ -103,122 +103,17 @@ export function LatencyDiagnostic({ latency }: LatencyDiagnosticProps) {
           </Text>
 
           <Flex alignItems="center">
-            <Flex
-              gap={1}
-              width="24px"
-              flexDirection="column"
-              alignItems="flex-start"
-              css={`
-                flex-grow: 0;
-              `}
-            >
+            <IconContainer text="You" alignItems="flex-start">
               <Icons.User />
-              <Text>You</Text>
-            </Flex>
-
-            <Flex
-              gap={1}
-              flexDirection="column"
-              alignItems="center"
-              css={`
-                flex-grow: 1;
-                position: relative;
-              `}
-            >
-              <Flex
-                gap={1}
-                flexDirection="row"
-                alignItems="center"
-                width="100%"
-                css={`
-                  padding: 0 16px;
-                `}
-              >
-                <Icons.ChevronLeft
-                  size="small"
-                  color="text.muted"
-                  css={`
-                    left: 8px;
-                    position: absolute;
-                  `}
-                />
-                <Line />
-                <Icons.ChevronRight
-                  size="small"
-                  color="text.muted"
-                  css={`
-                    right: 8px;
-                    position: absolute;
-                  `}
-                />
-              </Flex>
-              <Text color={colors.client}>{latency.client}ms</Text>
-            </Flex>
-
-            <Flex
-              gap={1}
-              width="24px"
-              flexDirection="column"
-              alignItems="center"
-              css={`
-                flex-grow: 0;
-              `}
-            >
-              <TeleportGearIcon size={24}></TeleportGearIcon>
-              <Text>Teleport</Text>
-            </Flex>
-
-            <Flex
-              gap={1}
-              flexDirection="column"
-              alignItems="center"
-              css={`
-                flex-grow: 1;
-                position: relative;
-              `}
-            >
-              <Flex
-                gap={1}
-                flexDirection="row"
-                alignItems="center"
-                width="100%"
-                css={`
-                  padding: 0 16px;
-                `}
-              >
-                <Icons.ChevronLeft
-                  size="small"
-                  color="text.muted"
-                  css={`
-                    left: 8px;
-                    position: absolute;
-                  `}
-                />
-                <Line />
-                <Icons.ChevronRight
-                  size="small"
-                  color="text.muted"
-                  css={`
-                    right: 8px;
-                    position: absolute;
-                  `}
-                />
-              </Flex>
-              <Text color={colors.server}>{latency.server}ms</Text>
-            </Flex>
-
-            <Flex
-              gap={1}
-              width="24px"
-              flexDirection="column"
-              alignItems="flex-end"
-              css={`
-                flex-grow: 0;
-              `}
-            >
+            </IconContainer>
+            <Leg color={colors.client} latency={latency.client} />
+            <IconContainer text="Teleport" alignItems="center">
+              <TeleportGearIcon size={24} />
+            </IconContainer>
+            <Leg color={colors.server} latency={latency.server} />
+            <IconContainer text="Server" alignItems="flex-end">
               <Icons.Server />
-              <Text>Server</Text>
-            </Flex>
+            </IconContainer>
           </Flex>
 
           <Flex flexDirection="column" alignItems="center">
@@ -246,6 +141,81 @@ export function LatencyDiagnostic({ latency }: LatencyDiagnosticProps) {
   );
 }
 
+type IconContainerProps = {
+  text: string;
+  alignItems: 'flex-start' | 'center' | 'flex-end';
+};
+
+const IconContainer: React.FC<IconContainerProps> = ({
+  children,
+  text,
+  alignItems,
+}) => (
+  <Flex
+    gap={1}
+    width="24px"
+    flexDirection="column"
+    alignItems={alignItems}
+    css={`
+      flex-grow: 0;
+    `}
+  >
+    {children}
+    <Text>{text}</Text>
+  </Flex>
+);
+
+type LegProps = {
+  color: latencyColor;
+  latency: number;
+};
+
+const Leg: React.FC<LegProps> = ({ color, latency }) => (
+  <Flex
+    gap={1}
+    flexDirection="column"
+    alignItems="center"
+    css={`
+      flex-grow: 1;
+      position: relative;
+    `}
+  >
+    <DoubleSidedArrow />
+    <Text color={color}>{latency}ms</Text>
+  </Flex>
+);
+
+// Looks like `<----->`
+const DoubleSidedArrow = () => {
+  return (
+    <Flex
+      gap={1}
+      flexDirection="row"
+      alignItems="center"
+      width="100%"
+      p="0 16px"
+    >
+      <Icons.ChevronLeft
+        size="small"
+        color="text.muted"
+        css={`
+          left: 8px;
+          position: absolute;
+        `}
+      />
+      <Line />
+      <Icons.ChevronRight
+        size="small"
+        color="text.muted"
+        css={`
+          right: 8px;
+          position: absolute;
+        `}
+      />
+    </Flex>
+  );
+};
+
 const Container = styled.div`
   background: ${props => props.theme.colors.levels.elevated};
   padding: ${props => props.theme.space[4]}px;
```

</details>

It ends up looking like

```tsx
export function LatencyDiagnostic({ latency }: LatencyDiagnosticProps) {
  const colors = latencyColors(latency.client, latency.server);

  return (
    <MenuIcon Icon={Icons.Wifi} buttonIconProps={{ color: colors.total }}>
      <Container>
        <Flex gap={5} flexDirection="column">
          <Text textAlign="left" typography="h3">
            Network Connection
          </Text>

          <Flex alignItems="center">
            <IconContainer text="You" alignItems="flex-start">
              <Icons.User />
            </IconContainer>
            <Leg color={colors.client} latency={latency.client} />
            <IconContainer text="Teleport" alignItems="center">
              <TeleportGearIcon size={24} />
            </IconContainer>
            <Leg color={colors.server} latency={latency.server} />
            <IconContainer text="Server" alignItems="flex-end">
              <Icons.Server />
            </IconContainer>
          </Flex>

          <Flex flexDirection="column" alignItems="center">
            <Flex gap={1} flexDirection="row" alignItems="center">
              {colors.total === latencyColor.error && (
                <Icons.WarningCircle size={20} color={colors.total} />
              )}

              {colors.total === latencyColor.warn && (
                <Icons.Warning size={20} color={colors.total} />
              )}

              {colors.total === latencyColor.ok && (
                <Icons.CircleCheck size={20} color={colors.total} />
              )}

              <Text bold fontSize={2} textAlign="center" color={colors.total}>
                Total Latency: {latency.total}ms
              </Text>
            </Flex>
          </Flex>
        </Flex>
      </Container>
    </MenuIcon>
  );
}
```

which is easier on the reader/maintainer.